lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            self.pkg = pkg
+            if not self.pkg.spec.concrete:
+                raise ValueError("{0} must have a concrete spec"
+                                 .format(self.pkg.unique_id))
+        else:
+            raise ValueError("{0} must be a package".format(str(pkg)))
+
+        # Initialize the status to an active state.  The status is used to
+        # ensure priority queue invariants when tasks are "removed" from the
+        # queue.
+        if status != STATUS_REMOVED:
+            self.status = status
+        else:
+            msg = "Cannot create a build task for {0} with status '{1}'"
+            raise InstallError(msg)
+
+        # Package is associated with a bootstrap compiler
+        self.compiler = compiler
+
+        # The initial start time for processing the spec.
+        # TODO: Should this be tied to the status?
+        self.start = start
+
+        # Number of times the task has been queued
+        self.attempts = attempts + 1
+
+        # Set of dependents
+        self.dependents = set(d.package.unique_id for d
+                              in self.spec.dependents())
+
+        # Set of dependencies
+        #
+        # Be consistent wrt use of dependents and dependencies.  That is,
+        # if use traverse for transitive dependencies, then must remove
+        # transitive dependents on failure.
+        self.dependencies = set(d.package.unique_id for d in
+                                self.spec.dependencies() if
+                                d.package.unique_id != self.pkg.unique_id)
+
+        # List of uninstalled dependencies, which is used to establish
+        # the priority of the build task.
+        #
+        self.uninstalled_deps = set(pkg_id for pkg_id in self.dependencies if
+                                    pkg_id not in installed)
+
+        # Ensure the task gets a unique sequence number to preserve the
+        # order in which it was added.
+        self.sequence = next(_counter)
+
+    def __str__(self):
+        """Returns a printable version of the build task."""
+        attempts = '#attempts={0}'.format(self.attempts)
+        dependents = '#dependents={0}'.format(len(self.dependents))
+        dependencies = '#dependencies={0}'.format(len(self.dependencies))
+        start = 'start={0}s'.format(self.start)
+        status = 'status={0}'.format(self.status)
+        return ('(pri={0}, seq={1}, id={2}, {3}, {4}, {5}, {6}, {7})'
+                .format(self.priority, self.sequence, self.pkg_id, dependents,
+                        dependencies, start, attempts, status))
+
+    def flag_installed(self, installed):
+        """
+        Ensure the dependency is not considered to still be uninstalled.
+
+        Args:
+            installed (list of str): the identifiers of packages that have
+                been installed so far
+        """
+        uninstalled = list(self.uninstalled_deps)
+        for pkg_id in uninstalled:
+            if pkg_id in installed:
```

Comment:
Minor preference towards using sets here to get the intersection of `installed` and `uninstalled`.

Reply:
Sure.  Changed.